### PR TITLE
Read invitation parameters from wiki page

### PR DESCRIPTION
**Pull request by BenignBeppe**
Max age and min edits to qualify for invitation are defined in
hb_config. Values can be read from wiki pages.

### CODE REVIEW

File: hb_queries.py [hunk 1]
```diff
@@ -36,22 +37,26 @@ def make_teahouse_experiment_newbies(self):
...
+
         query = u"""
 INSERT IGNORE INTO {table:s}
 (user_id, user_name, user_registration, user_editcount, sample_date)
 SELECT user_id, user_name, user_registration, user_editcount, NOW()
 FROM {wikidb:s}.user
 WHERE user_registration > DATE_FORMAT(
-    DATE_SUB(NOW(),INTERVAL 2 DAY),'%Y%m%d%H%i%s')
-AND user_editcount >=5
+    DATE_SUB(NOW(),INTERVAL {max_user_age} DAY),'%Y%m%d%H%i%s')
```
Comment (lokal-profil) [inline hb_queries.py]: do we want to force formatting on this and the next parameter?

File: page_reader.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    if page in pages:
+        # Read cached page, if available.
+        page_content = pages[page]
+    else:
+        response = requests.post(
+            hb_config.apiurl,
+            data={
+                'action': "query",
+                'titles': page,
```
Comment (lokal-profil) [inline page_reader.py:60]: I believe you might have to do page.encode('utf-8') here.

But you can check with a unicode page name to make sure.

File: page_reader.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+def _read_page(page):
+    """Read the content of a page, given its title.
+
+    The content will be cached the first time a page is read, allowing
+    multiple calls to this function for the same page without sending
+    multiple requests.
+    """
+
+    if page in pages:
```
Comment (lokal-profil) [inline page_reader.py]: You should standardise the pagename first (capitalize first letter and either all spaces as underscores or all underscores as pages) to ensure the cached one is always found.

File: page_reader.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            response_pages[response_pages.keys()[0]]["revisions"][0]["*"]
+        # Chache page content.
+        pages[page] = page_content
+    return page_content
+
+def read_list_page(page):
+    """Read a new line separated list from a page."""
+
+    page_content = _read_page(page)
+    list_ = [i for i in page_content.split("\n")]
```
Comment (lokal-profil) [inline page_reader.py]: Is the loop only there to get rid of the last empty line or do you also expect more empty lines?

Wondering since in the first scenario `page_content.strip().split("\n")` should work.

File: scripts/teahouse/send_th_invites.py [hunk 5]
```diff
@@ -18,17 +18,22 @@
...
+import random
+import re
+
 import hb_toolkit
 import hb_output_settings
 import hb_profiles
-import random
+import page_reader
+
+RE_USER_NAME = re.compile(r"\[\[User:(.*?)(\]\]|\|)")
```
Comment (lokal-profil) [inline scripts/teahouse/send_th_invites.py]: "User:" depends on the language of the wiki. E.g. sv.wiki you can write either "User:" or "Användare:". Sadly for us signatures default to using the local language version.

So we would need `RE_USER_NAME = re.compile(r"\[\[(User|Användare):(.*?)(\]\]|\|)")` here and `RE_USER_NAME.match(wikitext).group(2)` below.

Of course it would be better to load the local language user-namespace from the config. (We don't need to load the english one since that is always allowed)

File: scripts/teahouse/send_th_invites.py [hunk 6]
```diff
@@ -78,8 +93,10 @@ def inviteGuests(prof, message_text, inviter):
     else:
         pass
 #     inviters = params['inviters'] #for TWA
+    page_list = page_reader.read_list_page("User:FikarummetBot/Inviters")
```
Comment (lokal-profil) [inline scripts/teahouse/send_th_invites.py]: This pagename should be in the config (and sample).

I would even say that the config should just return a clean list here and deciding whether to load from page or from a provided list in the config (and possibly also any username from wikitext extraction) should be handled before.

File: scripts/teahouse/send_th_invites.py [hunk 7]
```diff
@@ -78,8 +93,10 @@ def inviteGuests(prof, message_text, inviter):
...
+    inviter_list = [user_name_from_wikitext(i) for i in page_list]
     inviters = getEligibleInviters(elig_check, params['inviters'])
```
Comment (lokal-profil) [inline scripts/teahouse/send_th_invites.py]: This should take `inviter_list` rather than `params['inviters']`

File: scripts/teahouse/send_th_invites.py [hunk 8]
```diff
@@ -18,17 +18,22 @@
...
+
+RE_USER_NAME = re.compile(r"\[\[User:(.*?)(\]\]|\|)")
 
 def getEligibleInviters(elig_check, potential_inviters):
     """Filter out inviters with no edits in the last 21 days."""
     eligible_inviters = [x for x in potential_inviters if elig_check.determineInviterEligibility(x, 21)]
     return eligible_inviters
 
-def getEligibleInvitees(elig_check, potential_invitees, skip_templates):
+def getEligibleInvitees(elig_check, potential_invitees):
```
Comment (lokal-profil) [inline scripts/teahouse/send_th_invites.py:34]: Might be worth adding a comment on line 79 so that we remember that we are in fact still sending `skip templates` (and `output namespace`) to the underlying function.

**Review by lokal-profil — CHANGES_REQUESTED** [review]
Good in general. A few minor changes though.

Also if you can drop the now unneeded entries from `hb_output_settings` it should be easier to spot which are left to be addressed.

File: hb_queries.py (hunk 1 above)
Comment (BenignBeppe) [reply hb_queries.py]: Seems like a good thing to check. I'll add this.

File: page_reader.py (hunk 4 above)
Comment (BenignBeppe) [reply page_reader.py]: I'm honestly not sure why I wrote it like this. Just `page_content.split("\n")` works, so I changed to this. Trailing new lines seems to be stripped before the content is received, as well.

File: scripts/teahouse/send_th_invites.py (hunk 6 above)
Comment (BenignBeppe) [reply scripts/teahouse/send_th_invites.py]: I moved the inviter list to `hb_config`.

File: scripts/teahouse/send_th_invites.py (hunk 7 above)
Comment (BenignBeppe) [reply scripts/teahouse/send_th_invites.py]: Done.

File: page_reader.py (hunk 2 above)
Comment (BenignBeppe) [reply page_reader.py:60]: Amazingly enough, character encoding doesn't seem to cause any trouble here. I did quick test and "User:FikarummetBot/Å" worked, both as `str` and `unicode`. I did changed the call to `get()´.

Comment (lokal-profil) [reply page_reader.py:60]: Great. I guess my worry stems from pre-`requests` days =)

File: scripts/teahouse/send_th_invites.py (hunk 5 above)
Comment (BenignBeppe) [reply scripts/teahouse/send_th_invites.py]: User namespace is now defined in `hb_config`. "User" is always valid.

File: page_reader.py (hunk 3 above)
Comment (BenignBeppe) [reply page_reader.py]: Added function for normalizing page title.

File: scripts/teahouse/send_th_invites.py (hunk 8 above)
Comment (BenignBeppe) [reply scripts/teahouse/send_th_invites.py:34]: Done.

**Comment by BenignBeppe** [comment]
I've commented out the parameters no longer used in `hb_output_settings`. Of the remaining, skipped templates and user talk name space should be relatively straight forward. 'select query' may take some investigation to figure out what it actually does.

File: page_reader.py [hunk 9]
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
         pages[page] = page_content
     return page_content
 
+def _normalize_title(title):
+    """Normalize a page title.
+
+    If the title contains a colon, it is assumed that it is a namspace
```
Comment (lokal-profil) [inline page_reader.py]: typo namespace

File: page_reader.py [hunk 10]
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
...
+    delimiter.
+    """
+
+    parts = title.partition(":")
+    if parts[1] != "":
```
Comment (lokal-profil) [inline page_reader.py]: You can just do if parts[1]

File: page_reader.py [hunk 11]
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
...
+        # Since a colon was found, this should contain a namespace.
+        namespace, _, name = parts
+        capitalized_name = name[0].upper() + name[1:]
```
Comment (lokal-profil) [inline page_reader.py]: You should also capitalize the namespace `user:lisa == `User:Lisa`

File: page_reader.py [hunk 12]
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
...
+        capitalized_title = "{}:{}".format(namespace, capitalized_name)
+    else:
+        name = parts[0]
+        capitalized_name = name[0].upper() + name[1:]
```
Comment (lokal-profil) [inline page_reader.py]: should be capitalized_title in order to get picked up by normalized_title below

File: page_reader.py [hunk 13]
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
...
+    normalized_title = capitalized_title.replace("_", " ")
+    return normalized_title
```
Comment (lokal-profil) [inline page_reader.py:103]: Could this (i.e from partition to return) not be done as
```
namespace, _, name = title.partition(":")
if name:
    namespace = namespace[0].upper() + namespace[1:]
else:
    name = namespace
    namespace = ''
name = name[0].upper() + name[1:]
capitalized_title = "{}{}{}".format(namespace, _, name)
normalized...
```

File: page_reader.py (hunk 12 above)
Comment (lokal-profil) [reply page_reader.py]: see comment below

File: page_reader.py (hunk 11 above)
Comment (lokal-profil) [reply page_reader.py]: see comment below

File: page_reader.py (hunk 10 above)
Comment (lokal-profil) [reply page_reader.py]: see comment below

File: scripts/teahouse/send_th_invites.py [hunk 14]
```diff
@@ -25,8 +25,13 @@
 import hb_output_settings
 import hb_profiles
 import page_reader
+import hb_config
 
-RE_USER_NAME = re.compile(r"\[\[User:(.*?)(\]\]|\|)")
+# Build an RE pattern that accepts None as user namespace to just use
+# the default "User".
+RE_USER_NAME = re.compile(r"\[\[(User{:s}):(.*?)(\]\]|\|)".format(
+    ("|" + str(hb_config.user_namespace)) * (hb_config.user_namespace != None)
```
Comment (lokal-profil) [inline scripts/teahouse/send_th_invites.py]: I believe "is not" is the more common test for None comparison

File: scripts/teahouse/send_th_invites.py [hunk 15]
```diff
@@ -71,11 +76,13 @@ def user_name_from_wikitext(wikitext):
       [[User:FikarummetBot|FikarummetBot]] ([[User talk:FikarummetBot|talk]]) 09:12, 17 October 2016 (UTC) -> FikarummetBot
     """
 
-    return RE_USER_NAME.match(wikitext).group(1)
+    return RE_USER_NAME.match(wikitext).group(2)
```
Comment (lokal-profil) [inline scripts/teahouse/send_th_invites.py]: I think we need a try/except block here since (when I quickly test it) I get AttributeError if the input is weirdly formatted. Since we are reading this from a wikipage I suggest adding the robustness of simply skipping any such. Easiest is to simply return `None` on (that) exception and add `inviter_list = filter(None, inviter_list)` later.

File: scripts/teahouse/send_th_invites.py [hunk 16]
```diff
@@ -93,9 +100,8 @@ def user_name_from_wikitext(wikitext):
     else:
         pass
 #     inviters = params['inviters'] #for TWA
-    page_list = page_reader.read_list_page("User:FikarummetBot/Inviters")
-    inviter_list = [user_name_from_wikitext(i) for i in page_list]
-    inviters = getEligibleInviters(elig_check, params['inviters'])
+    inviter_list = [user_name_from_wikitext(i) for i in hb_config.inviters]
```
Comment (lokal-profil) [inline scripts/teahouse/send_th_invites.py]: could you add this variable to the config sample file?

**Review by lokal-profil — CHANGES_REQUESTED** [review]
comments on _normalize_title and RE_USER_NAME

File: page_reader.py (hunk 13 above)
Comment (BenignBeppe) [reply page_reader.py:103]: Yep, that's probably easier to read. I added it with a couple of minor changes.

File: page_reader.py (hunk 9 above)
Comment (BenignBeppe) [reply page_reader.py]: Fixed.

File: scripts/teahouse/send_th_invites.py (hunk 14 above)
Comment (BenignBeppe) [reply scripts/teahouse/send_th_invites.py]: Done.

File: scripts/teahouse/send_th_invites.py (hunk 16 above)
Comment (BenignBeppe) [reply scripts/teahouse/send_th_invites.py]: Added.

File: scripts/teahouse/send_th_invites.py (hunk 15 above)
Comment (BenignBeppe) [reply scripts/teahouse/send_th_invites.py]: Now, only valid user names (from links) are added. When an entry doesn't match, a debug message is written to stdout.

File: hb_config.sample.py [hunk 17]
```diff
@@ -51,3 +51,7 @@
 
 # The summary for the message. {invitee:s} is the name of the invited user.
 summary = '{invitee:s}, you are invited to the Teahouse!'
+
+# The user that can be mentioned in `message` above. A user will be
+# selected at random from this list, for each message.
```
Comment (lokal-profil) [inline hb_config.sample.py]: Maybe mention page_reader.read_user_list_page() here.

**Review by lokal-profil — APPROVED** [review]
